optax/transforms/_freezing.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""utilites for freezing parameters"""
+
+from typing import Union
+import chex
+import jax
+import jax.numpy as jnp
+from optax._src import base
+from optax.transforms import masked, partition
+
+
+def _is_bool_scalar(x) -> bool:
+    """
+    Return True if x is a Python bool or a 0-d JAX boolean array.
+    """
+    return isinstance(x, bool) or (
+        isinstance(x, jnp.ndarray)
+        and x.shape == ()
+        and x.dtype == jnp.bool_
+    )
+
+
+def freeze(mask: Union[bool, chex.ArrayTree]) -> "base.GradientTransformation":
+    """
+    Create a transformation that zeros out gradient updates according to `mask`,
+    essentially freezing (i.e. holding constant) those parameters.
+
+    The mask can be:
+      • a single boolean (or 0-d JAX bool array), causing every parameter to be
+        either all-frozen (True) or all-trainable (False), or
+      • a PyTree of booleans matching the structure of the parameters, where
+        each leaf indicates whether that specific parameter leaf should be
+        frozen (True) or left unchanged (False).
+
+    Returns:
+        An Optax `GradientTransformation` which applies `set_to_zero()` wherever
+        `mask==True`, and leaves other gradients intact.
+    """
+    def mask_fn(params: base.PyTree) -> base.PyTree:
+        if _is_bool_scalar(mask):
+            return jax.tree.map(lambda _: bool(mask), params)

```

Comment:
This will likely throw a concretization error if mask is a tracer.

Because tree.map supports a prefix map, you can rely on jax.tree.map directly in both scalar and nested case.

Does `return masked(base.set_to_zero(), mask)` work in all your test cases? (Instead of creating a `mask_fn`)